portfolio/src/main/java/com/google/sps/data/HighestPriorityFirstScheduler.java

```diff
@@ -0,0 +1,118 @@
+      Comparator.comparing(ScheduledTask::getStartTime);
+
+  /** This method schedules tasks from highest to lowest priority and */
+  public Collection<ScheduledTask> schedule(
+      Collection<CalendarEvent> events,
+      Collection<Task> tasks,
+      Instant workHoursStartTime,
+      Instant workHoursEndTime) {
+
+    List<CalendarEvent> eventsList = new ArrayList<CalendarEvent>(events);
+    List<Task> tasksList = new ArrayList<Task>(tasks);
+
+    TaskQueue taskQueue = new TaskQueue(tasksList, getSchedulingAlgorithmType());
+
+    CalendarEventsGroup calendarEventsGroup =
+        new CalendarEventsGroup(eventsList, workHoursStartTime, workHoursEndTime);
+
+    List<TimeRange> availableTimes = calendarEventsGroup.getFreeTimeRanges();
+    TimeRangeGroup availableTimesGroup = new ArrayListTimeRangeGroup(availableTimes);
+    List<TimeRange> currentAvailableTimes = getAvailableTimeRangesList(availableTimesGroup);
+
+    List<ScheduledTask> scheduledTasks = new ArrayList<ScheduledTask>();
+
+    int availableTimesIndex = 0;
+
+    Instant currentScheduleTime = workHoursStartTime;
+
+    while (availableTimesIndex < currentAvailableTimes.size() && !taskQueue.isEmpty()) {
+      TimeRange availableTimeRange = currentAvailableTimes.get(availableTimesIndex);
+      Task task = taskQueue.peek();
+
+      if (availableTimeRange.start().isAfter(currentScheduleTime)) {
+        currentScheduleTime = availableTimeRange.start();
+      }
+
+      if (!currentScheduleTime
+          .plusSeconds(task.getDuration().getSeconds())
+          .isAfter(availableTimeRange.end())) {
+        ScheduledTask scheduledTask =
+            new ScheduledTask(task, currentScheduleTime, Optional.of(true));
+        scheduledTasks.add(scheduledTask);
+
+        // Delete the TimeRange that has been scheduled over so that different priority tasks
+        // won't be scheduled over the same time.
+        availableTimesGroup.deleteTimeRange(
+            TimeRange.fromStartEnd(
+                scheduledTask.getStartTime(),
+                scheduledTask
+                    .getStartTime()
+                    .plusSeconds(scheduledTask.getTask().getDuration().getSeconds())));
+        // After deleting a TimeRange, we must renew the list.
+        currentAvailableTimes = getAvailableTimeRangesList(availableTimesGroup);
+
+        // Push the currentScheduleTime back by the amount that was scheduled.
+        currentScheduleTime = currentScheduleTime.plusSeconds(task.getDuration().getSeconds());
+        taskQueue.remove();
+
+        // If the next task's priority is different from the task that was just scheduled,
+        // then reset the availableTimesIndex, currentScheduleTime, and make the loop
+        // grab a TimeRange. We go back to the first availableTimeRange in order to schedule
+        // more tasks towards the beginning of the availableTimes.
+        if (isNextTaskDifferentPriority(taskQueue, task)) {
```

Comment:
I'm not sure why this is necessary. But I understand that you want to make use of earlier, available time ranges. I think the following approach might be simpler.

1. First, you order the tasks by their priority (in descending order).
2. For each task, you identify the first time range that can fit that task. And schedule it there.

I think the current implementation is somewhat overly complex. The suggested approach above should be simpler, I think.

Reply:
The reasoning behind this is that since the tasks are also ordered from shortest to longest, when we're sorting tasks with the same priority we can sort all the possible tasks for a priority in one run through of the availableTimeRanges instead of possibly (worst case scenario)having to iterate to the last availableTimeRange to schedule each task.

Reply:
Okay. Feel free to keep this logic.

Reply:
I agree with James here, although I think it's fine to leave it as is now.

 The only thing is that with this logic I think you get results that are a little surprising. When I think "Highest Priority FIrst", I tend to think that it will try to schedule the highest priority tasks as soon as possible, but what's really happening (since you're also sorting the free time ranges by length) is that it's *choosing* the timeslot for the highest priority first, but not trying to put it into the first timeslot.

For example, if I have five 1-hour tasks, and I have a 5 hour block open at the beginning of the day and two 1-hour blocks later, I'll actually end up working on the 3rd, 4th, and 5th highest priority tasks first, then later work on the 1st and 2nd highest priority. Do I have that right? Is that what we're going for with this algorithm?

Reply:
I am not sorting free time ranges by length. The only sorting that is done is sorting the Tasks by priority, then by duration and sorting the scheduledTasks by start time before returning them.

Reply:
I see, sorry I misunderstood your comment above but that makes sense.